docs/spec/indexing.rst

```diff
@@ -0,0 +1,144 @@
+
+    index.as_flat_series()
+
+
+Partition Indices
+-----------------
+
+The first index type kartothek offers is a partition index. The partition index is created by partitioning a dataset in a hive-like partition scheme.
+
+.. ipython:: python
+    :suppress:
+
+    import string
+    import pandas as pd
+    from kartothek.core.utils import ensure_store
+
+    store = ensure_store("hmemory://")
+    from kartothek.io.eager import store_dataframes_as_dataset
+
+.. ipython:: python
+
+    df = pd.DataFrame(
+        {
+            "PartField": ["A"] * 5 + ["B"] * 5,
+            "IndexedField": list(range(5)) + list(range(3, 8)),
+            "Payload": [string.ascii_letters[i] for i in range(10)],
+        }
+    )
+    dm = store_dataframes_as_dataset(
+        store=store,
+        dataset_uuid="indexing_docs",
+        dfs=[df],
+        partition_on=["PartField"],
+        secondary_indices=["IndexedField"],
+    ).load_all_indices(store)
+
+    sorted(store.keys())
+
+    part_index = dm.indices["PartField"]
+    part_index
+
+This kind of index is also called a `primary index`. This implies the property that a given file is guaranteed to only contain **one** unique value of the given field. This can also be observed when investigating the flat structure of the index.
+
+.. ipython:: python
+
+    part_index.as_flat_series()
+
+This property makes this kind of index very powerful if used correctly since it prunes the partitions exactly to the user query and enables exact removal of data when mutating datasets (see :doc:`../guide/mutating_datasets`).
+
+For data with high cardinality this kind of index is not well suited since it would result in a highly fragmented dataset with too many, too small files.
+
+
+Secondary indices
+-----------------
+
+Secondary indices are the most powerful type of indices which allow us to reference files without having to encode any kind of values in the keys. They can be created by supplying the `secondary_indices` keyword argument as shown above. The user interaction works similarly to the
+
+
+Persistence
+~~~~~~~~~~~
+
+A secondary index is persisted as a Parquet file with the following
+(Parquet) schema:
+The field name corresponds to the name of the column in the persisted
+DataFrame.
+The partition is a list of partition identifiers, as used in the keys of
+the partitions map and the data filename. (Note: the partition identifier
+is used instead of the data filename as a single partition can span multiple
+files containing different column sets using the same row selection.)
+
+
+Typing
+------
+
+Every index has a well defined arrow data type which is usually inferred automatically and ensured to be consistent with the overall dataset schema.
+
+.. ipython:: python
+
+    part_index.dtype
+
+Kartothek generally does not impose any restrictions on the data types to be used as long as they are compatible and in alignment with the :doc:`pyarrow pandas integration<python/pandas>`.
```

Comment:
It is true for *payload data* that kartothek essentially supports what arrow does, but for indices, it's a bit more complex, as there is string-to-value conversion involved, so there is actually a limited set supported by kartothek:
https://github.com/JDASoftwareGroup/kartothek/blob/5b239bd906b8e01fe1691a26745330b018814925/kartothek/core/index.py#L144
... which ends with
https://github.com/JDASoftwareGroup/kartothek/blob/5b239bd906b8e01fe1691a26745330b018814925/kartothek/core/index.py#L201

Therefore, for indices, I think this statement is misleading.

BTW: does the documentation  already contain the statement about what data types are supported for payload data? I could not find it ...

Reply:
> BTW: does the documentation already contain the statement about what data types are supported for payload data? I could not find it ...


The documentation includes a very long description about what types are used and how they are supported https://kartothek.readthedocs.io/en/stable/spec/type_system.html

If there is something missing and/or wrong, I'd rather deal with it by fixing it over there instead of documenting it again. Otherwise I moved this statement and put it into https://kartothek.readthedocs.io/en/stable/spec/serialization.html with a link to the type system page